### PR TITLE
RFC: Avoid running test suite with doc only changes

### PR DESCRIPTION
**Pull request by omus**
I got the idea from a comment in Slack from @fredrikekre.

### CODE REVIEW

**Comment by omus** [comment]
I may be able to get this working on Circle CI via `git diff` and getting a commit range using: https://discuss.circleci.com/t/get-list-of-commits-in-build/15725/2

**Comment by fredrikekre** [comment]
Great! It would be nice if we could also detect docstring-only changes, but I suspect that is more difficult.

**Comment by omus** [comment]
Looks like I have AppVeyor working how I want now with only changes to the "doc" folder. I'll see if I can add back the "message" feature.

**Comment by omus** [comment]
Travis CI seems to be working and cut the runtime in half:

Running test suite
```
Linux 32-bit: 2 hrs 2 min 20 sec
Linux 64-bit: 2 hrs 4 min 59 sec
macOS:        1 hr 49 min 41 sec
```
Skipping test suite:
```
Linux 32-bit: 50 min 24 sec
Linux 64-bit: 1 hr 3 min 24 sec
macOS:        42 min
```

File: appveyor.yml [hunk 1]
```diff
@@ -12,10 +12,15 @@ branches:
...
     - master
     - /^release-.*/
+    - /^cv.*/
 
+# Note: use `[ci skip]` or `[skip ci]` in the commit message title and AppVeyor won't be
+# built for that commit. You can use `[skip appveyor]` to explicitly skip AppVeyor and
+# allow other CI to still run.
 skip_commits:
-# Add [av skip] to commit messages for docfixes, etc to reduce load on queue
-  message: /\[av skip\]/
```
Comment (omus) [inline appveyor.yml]: I need to run another test to validate this removal is necessary as originally I messed up the `files` rule.

File: .travis.yml [hunk 2]
```diff
@@ -139,8 +140,10 @@ script:
         /tmp/julia/bin/julia-debug --sysimage-native-code=no -e 'true'
     - /tmp/julia/bin/julia -e 'versioninfo()'
     - pushd /tmp/julia/share/julia/test
-    - /tmp/julia/bin/julia --check-bounds=yes runtests.jl $TESTSTORUN &&
-      /tmp/julia/bin/julia --check-bounds=yes runtests.jl LibGit2/online OldPkg/pkg Pkg/pkg download
+    # skip tests if only files within the "doc" dir have changed
+    - if [ ! -z $(git diff --name-only $TRAVIS_COMMIT_RANGE | grep -v '^doc/') ]; then
```
Comment (omus) [inline .travis.yml]: Note: `git diff --name-only` always shows the file path relative to the root of the directory even when you are within a subdirectory.

File: appveyor.yml [hunk 3]
```diff
@@ -12,10 +12,17 @@ branches:
...
 
+# Note: use `[ci skip]` or `[skip ci]` in the commit message title and AppVeyor won't be
+# built for that commit. You can use `[skip appveyor]` to explicitly skip AppVeyor and
+# allow other CI to still run.
 skip_commits:
 # Add [av skip] to commit messages for docfixes, etc to reduce load on queue
   message: /\[av skip\]/
+# Skip running CI for changes only to the documentation
+  files:
+    - doc/
```
Comment (tkelman) [inline appveyor.yml]: skipping entirely doesn't seem like a good idea, doc-build failures that are platform specific would never get caught until after a merge

Comment (omus) [reply appveyor.yml]: Thanks for pointing that out. Do you have an example of a platform specific doc-build failure which would be caused by a doc change?

**Comment by omus** [comment]
Circle CI wasn't working correctly since I was performing the `git diff` in the wrong directory. Some of the code has been revised (including on Travis CI) to store the listing of files changed in the commit range within `FILES_CHANGED`.

**Comment by omus** [comment]
I noticed that sometimes the commit range includes a base commit which is unknown to the newly cloned repo. This seems to occur when a force push is involved:
```bash
$ git diff --name-only 2b44222c1d65...b9835b6ac332 --
fatal: bad revision '2b44222c1d65...b9835b6ac332'
```
Looking at my development repo you can see that `2b44222c1d65` is the original commit before the force push:
```
* b9835b6ac3 - (upstream/cv/doc-readme-link, cv/doc-readme-link) Fix doc README link (4 hours ago) <Curtis Vogt>
* f7707ad840 - (upstream/cv/ci-doc-speedup, cv/ci-doc-speedup) Compute files changed earlier [ci skip] (4 hours ago) <Curtis Vogt>
| * 2b44222c1d - Fix doc README link (2 days ago) <Curtis Vogt>
|/
* 0694191459 - fixup! Remove AppVeyor skip_commit for messages [ci skip] (2 days ago) <Curtis Vogt>
```
As a work around I now am using `git ls-files` as the fallback for when this occurs.

**Comment by omus** [comment]
Falling back to `git ls-files` introduced an issue which resulted in the error:
```
/home/travis/.travis/job_stages: line 57: [: too many arguments
```
The solution ended up being to use `-c` with grep to avoid processing to much data with bash's `-z` conditional.

File: appveyor.yml (hunk 3 above)
Comment (tkelman) [reply appveyor.yml]: I don't have the time available to dig through appveyor logs and find a specific example, but it has happened before. Documentation of platform-specific functionality will need the doc build to get run on each platform. And there have been issues in Documenter's handling of Windows paths that have come up.

**Comment by omus** [comment]
These changes should be ready to go now. I didn't get around to having the FreeBSD CI skip the doctests but since that CI is so fast it isn't really an issue.

**Comment by omus** [comment]
We may want to revise how AppVeyor is being handled (see https://github.com/JuliaLang/julia/pull/27356#discussion_r192536676).

File: appveyor.yml (hunk 3 above)
Comment (omus) [reply appveyor.yml]: Ok. If there is a doc-build issue that is specific to Windows I think that is something that Documenter should be testing for. In the event that the Julia documentation triggers a new Documenter issue specific to Windows we can address it at that time. I don't think it's the Julia CI's responsibility to test Documenter.

Comment (tkelman) [reply appveyor.yml]: It is the Julia CI's responsibility to make sure changes don't break the build, before the get merged. You'll miss platform-specific instances of that if you go ahead with this change as-is.

Comment (omus) [reply appveyor.yml]: To play it safe I'll disable this check and try to implement a conditional block like Travis CI in a future PR.

**Comment by omus** [comment]
I've disabled the AppVeyor skipping for now to avoid any missing any doc-build issues on Windows. In a future PR I hope to add test only skipping to AppVeyor and FreeBSD. Unfortunately I don't have the time at the moment so I'll get the non-controversial parts of this PR merged.

**Comment by omus** [comment]
I'll merge this later today

File: appveyor.yml [hunk 4]
```diff
@@ -13,9 +13,16 @@ branches:
     - master
     - /^release-.*/
 
+# Note: use `[ci skip]` or `[skip ci]` in the commit message title and AppVeyor won't be
```
Comment (tkelman) [inline appveyor.yml]: doesn't have to be in the title, can be anywhere in the message

**Comment by omus** [comment]
CI is fully passing. I'll make a slight adjustment to a comment and merge this right away.

**Pull request merged**